src/components/Trading/TradingForm.tsx

```diff
@@ -53,22 +58,88 @@ interface Props {
 
 function TradingForm(props: Props) {
   const isSmallScreen = useIsMobile()
-  const [primaryAsset, setPrimaryAsset] = React.useState<Asset | undefined>(props.initialPrimaryAsset)
-  const [primaryAmountString, setPrimaryAmountString] = React.useState("")
-  const [secondaryAsset, setSecondaryAsset] = React.useState<Asset>(Asset.native())
-  const [manualPrice, setManualPrice] = React.useState<ManualPrice>({})
-  const [priceMode, setPriceMode] = React.useState<"primary" | "secondary">("secondary")
+
+  const formik = useFormik<TradingFormValues>({
+    initialValues: {
+      primaryAsset: props.initialPrimaryAsset,
+      primaryAmountString: "",
+      secondaryAsset: Asset.native(),
+      manualPrice: undefined,
+      priceMode: "secondary"
+    },
+    validate(values) {
```

Comment:
We should not inline this validation function, but rather declare it before the component. Should be easy and reduces the component's size.

Reply:
👆
This comment still needs to be addressed 😉

Reply:
It's actually not that easy because the validation function uses other variables like the `secondaryAmount` which is calculated inside the `TradingForm` component and the calculation requires things like a `tradePair` which is accessed with a hook. So we cannot calculate the `secondaryAmount` in the validation function if it's outside the functional component since we can't get the tradepair. 

Do you have any ideas on how to circumvent this? Maybe I am just missing something.

Reply:
Hmm, true. One solution could be currying:

```ts
const prepareValidateFormFields = (tradePair, secondaryAmount) => (values) => {
  // ...
}

// ...

const formik = useFormik<TradingFormValues>({
  // ...
  validate: prepareValidateFormFields(tradePair, secondaryAmount)
})
```

Another solution would be to move the validation method into a hook of its own, so it can use other hooks or we pass those secondary variables when we call the hook (`const validate = useTradingFormValidation(tradePair, secondaryAmount)`).

I think I like the hook-based approach better, but we can talk about it tomorrow.